Extend LAPACKFullMatrix and Vector to keep the data upon resizing and add rank-1 update

davydden opened the pull request:
The idea is to keep the old matrix/vector upon resizing. If the size is bigger, you fill with zeroes. If it's smaller, take the upper (left) block. This is useful when writing solver algorithms that work with reduced hessians.

davydden commented:
./run-tests

davydden commented:
```
$ ctest -R "lapack"
100% tests passed, 0 tests failed out of 72
```

davydden commented:
@dealii/dealii anyone wants to review this PR? I would like to go on and use these new features...

kronbichler commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -141,6 +141,18 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   * matrix in which case the error will be thrown.
+   */
+  void add(const number a,
+           const Vector<number> &v);
```
I would give this a more verbose name, say `rank1_update`. We already use `add` for too many other things. Or do we already use it somewhere?

davydden replied:
no, we don't have anywhere `A = A + v v`. I am fine with renaming it to `rank1_update` if you find it more descriptive.

kronbichler commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -157,6 +169,33 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   * \mathbf A_{21} & \mathbf A_{22}
+   * \end{array}
+   * \right)
+   * \f]
```
Is there any aesthetic reason to make this second case a `leftarrow`? I would have found putting `A_{11}` on the right with a `rightarrow` more intuitive...

davydden replied:
will update.

kronbichler commented on include/deal.II/lac/utilities.h:
```diff
@@ -33,6 +36,67 @@ namespace Utilities
+     * @author Denis Davydov, 2017
+     */
+    template<typename NumberType>
+    std::array<NumberType,3> Givens_rotation(const NumberType &x,
```
Do we capitalize proper names in function interfaces elsewhere?

davydden replied:
in this file  we already have `Lanczos_largest_eigenvalue` and `Chebyshev_filter`, so I would say yes.

kronbichler replied:
Alright, I didn't check those, then I guess we keep those names.

bangerth replied:
We're not consistent. For example, we also use `householder` and `gauss_jordan` as lower-case words. I think we used to always use lower-case, but that train has left the station.

davydden replied:
@kronbichler @bangerth actually, I added `Lanczos_largest_eigenvalue` and `Chebyshev_filter` after the latest release (8.5), sometime in August. So I can easily rename all those to be lower-case.

Do we capitalize elsewhere?

bangerth replied:
I think we've always used lower-case for function names, even for proper names.

davydden replied:
last commit fixes this.

kronbichler commented on include/deal.II/lac/utilities.h:
```diff
@@ -144,6 +208,95 @@ namespace Utilities
+      AssertThrow(false, ExcNotImplemented());
+      std::array<NumberType,3> res;
+      return res;
+    }
```
Would you mind sorting the functions in terms of `NumberType` and `std::complex<NumberType>` as they appear in the declaration? While scrolling through it I missed the Givens rotations.

davydden replied:
done.

kronbichler commented on include/deal.II/lac/utilities.h:
```diff
@@ -144,6 +208,95 @@ namespace Utilities
+                                             const NumberType &g)
+    {
+      std::array<NumberType,3> res;
+      // naieve calculation for "r" may overflow or underflow:
```
Let us say `naive` here rather than trying to typeset the diaeresis with ascii characters

kronbichler commented on include/deal.II/lac/utilities.h:
```diff
@@ -144,6 +208,95 @@ namespace Utilities
+      // See Golub 2013, Matrix computations, Chapter 5.1.8
+      // Algorithm 5.1.3
+      // and
+      // Anderson (2000), Discontinuous Plane Rotations and the Symmetric Eigenvalue Problem. LAPACK Working Note 150, University of Tennessee, UT-CS-00-454, December 4, 2000.
```
Please break line at 80 characters

kronbichler commented on include/deal.II/lac/vector.h:
```diff
@@ -258,6 +258,32 @@ class Vector : public Subscriptor
+   */
+  void reinit_preserve (const size_type N);
+
+
```
only 1 empty line

kronbichler commented on include/deal.II/lac/vector.h:
```diff
@@ -258,6 +258,32 @@ class Vector : public Subscriptor
+   * \f]
+   * whereas if the desired size is smaller, then
+   * \f[
+   * \mathbf V_1 \leftarrow
```
same as above

kronbichler commented on include/deal.II/lac/vector.h:
```diff
@@ -258,6 +258,32 @@ class Vector : public Subscriptor
+   * \mathbf V_1   \\
+   * \mathbf V_2
+   * \end{array}
+   * \right)
```
We already have the second case if the vector is shrinked...

davydden replied:
that's true, but it's only the case if we have `omit_zeroing=false`, whereas I can not _encode_ the first one with the same function `reinit(size, bool)`. So I decided for a new function instead and then used the same name in LAPACKFullMatrix

kronbichler replied:
Sure, we do have that, I was more thinking about adding some more description pointing out this analogy in case someone wonders.

kronbichler commented on include/deal.II/lac/vector.templates.h:
```diff
@@ -986,11 +1023,14 @@ Vector<Number>::memory_consumption () const
   Utilities::System::posix_memalign ((void **)&new_values, 64, sizeof(Number)*max_vec_size);
+  // copy:
+  for (size_type i = 0; i < copy_n_el; ++i)
+    new_values[i] = values[i];
```
Should we use the old TBB partitioner and some of the internal vector operations to do this?

davydden replied:
i was concerned with this operation for dense linear algebra where vector size is `[1,20)` or so, so in this case it's not crucial. I would prefer to leave it for further PRs if/when someone applies it to larger vectors.

kronbichler replied:
Yes sure.

kronbichler commented on source/lac/lapack_full_matrix.cc:
```diff
@@ -193,6 +213,138 @@ LAPACKFullMatrix<number>::add (const number              a,
+  }
+}
+
+
```
Three empty lines here...

davydden replied:
done.

kronbichler commented on source/lac/lapack_full_matrix.cc:
```diff
@@ -193,6 +213,138 @@ LAPACKFullMatrix<number>::add (const number              a,
+
+
+
+
```
delete one empty line here

kronbichler commented on tests/lac/utilities_04.output:
```diff
@@ -0,0 +1,7 @@
+DEAL::0.00000
+DEAL::0.00000
+DEAL::0.00000
+DEAL::0.00000
```
I have not tried to figure out the Givens rotations for the particular values you give but is it intended that all results are zero?

davydden replied:
yes, it checks that `Rotation * [x y]^T = [r 0]^T` and then calculates residual of this equation.

kronbichler replied:
Ah, now I see, then this makes perfect sense. I admit I didn't read carefully enough. Maybe help future readers of the test case by adding the word `residual` to the output? (But that may be just my taste, I'm equally fine if you leave things as they are.)

davydden replied:
done.

kronbichler commented on tests/lac/utilities_05.output:
```diff
@@ -0,0 +1,6 @@
+DEAL::0.00000
+DEAL::0.00000
+DEAL::0.00000
+DEAL::0.00000
```
Same here.

davydden replied:
ditto, but for hyperbolic matrix.

kronbichler commented on tests/lapack/full_matrix_26.output:
```diff
@@ -0,0 +1,5 @@
+DEAL::OK
+DEAL::OK
+DEAL::OK
+DEAL::OK
```
Nice test!

kronbichler commented on tests/lapack/full_matrix_27.cc:
```diff
@@ -0,0 +1,80 @@
+                    ExcMessage("diff="+ std::to_string(diff)));
+      }
+
+  deallog << "Ok" << std::endl;
```
Nit-picking: You used `OK` in the other test...

kronbichler commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -157,6 +169,33 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   * \right)
+   * \f]
+   */
+  void reinit_preserve (const size_type size);
```
Does anyone have a preference regarding the name? I would have preferred `reinit_preserving_entries` or something equally verbose.

davydden replied:
I am fine with renaming

bangerth replied:
`grow_or_shrink()`?

kronbichler replied:
That's a good idea @bangerth, I like `grow_or_shrink` a lot. What do you think @davydden?

davydden replied:
since both of you like it, i will update 👍

davydden replied:
done.

davydden commented:
@kronbichler thanks for reviewing at this late hour 😄

davydden commented:
@kronbichler i think I addressed all your comments. Looks like I have not broken anything while doing so:

```
ctest -R "lac/utilities_04|lac/utilities_05|lapack/full_matrix_26|lapack/full_matrix_27|lapack/full_matrix_28|lapack/full_matrix_29|vector/vector_58"
...
100% tests passed, 0 tests failed out of 14
```

davydden commented:
@kronbichler last two squashed.

davydden commented:
failing test is clearly unrelated:
```
The following tests FAILED:
	589 - opencascade/arclength_boundary_02.debug (Failed)
Errors while running CTest
```

bangerth commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -141,6 +141,18 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   *
+   * This function also works for Cholesky factorization. Updating ($a>0$) is
+   * performed via Givens rotations, whereas downdating ($a<0$) via hyperbolic
+   * rotations. Note that the later case might lead to a negative definite
```
-> latter

bangerth commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -141,6 +141,18 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   * This function also works for Cholesky factorization. Updating ($a>0$) is
+   * performed via Givens rotations, whereas downdating ($a<0$) via hyperbolic
+   * rotations. Note that the later case might lead to a negative definite
+   * matrix in which case the error will be thrown.
```
-> an error will be thrown (because Cholesky factorizations are only valid for symmetric and <i>positive definite</i> matrices)

bangerth commented on include/deal.II/lac/lapack_full_matrix.h:
```diff
@@ -141,6 +141,18 @@ class LAPACKFullMatrix : public TransposeTable<number>
+   * Perform a rank-1 update of a symmetric matrix
+   * $ A \leftarrow A + a \, \rm v \rm v^T $.
+   *
+   * This function also works for Cholesky factorization. Updating ($a>0$) is
```
Is the description of the up/downdating procedure only applicable if the matrix actually represents a Cholesky factorization?

If so, maybe start the second sentence with "In that case, updating ..."

davydden replied:
correct, will update.

bangerth commented on include/deal.II/lac/utilities.h:
```diff
@@ -33,6 +36,67 @@ namespace Utilities
   {
 
+    /**
+     * Return elements of continuous Givens rotation matrix and norm of the input vector.
```
-> Return the elements of a continuous Givens rotation matrix and the norm of the input vector.

bangerth commented on include/deal.II/lac/utilities.h:
```diff
@@ -33,6 +36,67 @@ namespace Utilities
+                                             const NumberType &y);
+
+    /**
+     * Return elements of hyperbolic rotation matrix.
```
-> Return the elements of a hyperbolic rotation matrix.

davydden commented:
@bangerth I addressed the comments and squashed with the last commit.

bangerth commented:
OK. I'm good with merging. @kronbichler -- did you have anything else?
